File: IAIso-v5.0/core/iaiso-python/iaiso/middleware/langchain.py
"""LangChain integration — a callback handler that reports token and tool-call
events into a BoundedExecution.

Usage:
    from langchain_anthropic import ChatAnthropic
    from iaiso import BoundedExecution
    from iaiso.middleware.langchain import IAIsoCallbackHandler

    with BoundedExecution.start() as exec:
        handler = IAIsoCallbackHandler(exec)
        llm = ChatAnthropic(model="claude-opus-4-7", callbacks=[handler])
        result = llm.invoke("hello")

The handler intentionally does not abort runs on escalation — LangChain callbacks
are advisory. For enforcement, check `execution.check()` between steps in your
chain/agent logic, or wrap the underlying LLM with the provider-specific
middleware in `iaiso.middleware.anthropic` / `iaiso.middleware.openai`.

Install:  pip install iaiso[langchain]
"""

from __future__ import annotations

from typing import Any
from uuid import UUID

try:
    from langchain_core.callbacks import BaseCallbackHandler
    from langchain_core.messages import BaseMessage
    from langchain_core.outputs import LLMResult
except ImportError as exc:  # pragma: no cover
    raise ImportError(
        "langchain-core is required for iaiso.middleware.langchain. "
        "Install with `pip install iaiso[langchain]`."
    ) from exc

from iaiso.core import BoundedExecution


class IAIsoCallbackHandler(BaseCallbackHandler):
    """LangChain callback handler that accounts for LLM calls and tool uses.

    This handler does not block or raise on escalation. It records events into
    the attached BoundedExecution and emits audit events. Enforcement decisions
    (pause, reset, reject) must be made by the caller between chain steps.
    """

    def __init__(self, execution: BoundedExecution) -> None:
        self._execution = execution

    def on_llm_end(
        self,
        response: LLMResult,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        tokens = 0
        usage = (response.llm_output or {}).get("token_usage") or {}
        if isinstance(usage, dict):
            tokens = int(
                usage.get("total_tokens")
                or (usage.get("prompt_tokens", 0)
                    + usage.get("completion_tokens", 0))
            )
        self._execution.record_tokens(tokens=tokens, tag=f"langchain.llm:{run_id}")

    def on_tool_end(
        self,
        output: Any,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        self._execution.record_tool_call(
            name=f"langchain.tool:{run_id}", count=1
        )

    def on_chain_end(
        self,
        outputs: dict[str, Any],
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        # Chain boundaries increment depth accounting by 1 step with depth=1.
        # This lets the pressure model distinguish deep nested plans from
        # shallow ones even when total token count is similar.
        self._execution.record_step(depth=1, tag=f"langchain.chain:{run_id}")

File: IAIso-v5.0/core/iaiso-node/src/middleware/langchain.ts
/**
 * LangChain callback handler — hooks pressure accounting into a
 * LangChain runnable/chain via the `@langchain/core` callback API.
 *
 * Install:
 *   npm install @langchain/core
 *
 * Example:
 *   import { ChatAnthropic } from "@langchain/anthropic";
 *   import { BoundedExecution, IaisoCallbackHandler } from "@iaiso/core";
 *
 *   await BoundedExecution.run({ config, audit_sink }, async (exec) => {
 *     const handler = new IaisoCallbackHandler(exec);
 *     const model = new ChatAnthropic({ model: "claude-opus-4-7" });
 *     const response = await model.invoke(prompt, { callbacks: [handler] });
 *   });
 *
 * Tokens come from `llmOutput.tokenUsage` where available (Anthropic,
 * OpenAI). Tool-call accounting happens via `handleToolStart`.
 */

import { BoundedExecution } from "../core/execution.js";

interface TokenUsage {
  totalTokens?: number;
  promptTokens?: number;
  completionTokens?: number;
  inputTokens?: number;
  outputTokens?: number;
}

interface LLMResult {
  llmOutput?: {
    tokenUsage?: TokenUsage;
    usage?: TokenUsage;
  };
  generations?: Array<Array<{ text?: string; message?: unknown }>>;
}

/**
 * IAIso callback handler. Implements a minimal subset of the
 * `@langchain/core/callbacks` `BaseCallbackHandler` interface — just the
 * hooks we need (LLM-end + tool-start). Typed structurally so we don't
 * have to take `@langchain/core` as a hard import.
 */
export class IaisoCallbackHandler {
  readonly name = "iaiso_callback_handler";
  readonly _execution: BoundedExecution;
  readonly awaitHandlers = true;
  readonly ignoreLLM = false;
  readonly ignoreChain = false;
  readonly ignoreAgent = false;
  readonly ignoreRetriever = false;
  readonly ignoreCustomEvent = false;

  constructor(execution: BoundedExecution) {
    this._execution = execution;
  }

  async handleLLMEnd(output: LLMResult): Promise<void> {
    const usage = output.llmOutput?.tokenUsage ?? output.llmOutput?.usage ?? {};
    const tokens =
      Number(usage.totalTokens ?? 0) ||
      Number(usage.promptTokens ?? 0) + Number(usage.completionTokens ?? 0) ||
      Number(usage.inputTokens ?? 0) + Number(usage.outputTokens ?? 0);

    if (tokens > 0) {
      this._execution.recordTokens(tokens, "langchain.llm_end");
    }
  }

  async handleToolStart(
    tool: { name?: string; id?: string[] } | unknown,
  ): Promise<void> {
    const toolName =
      typeof tool === "object" && tool !== null && "name" in tool
        ? String((tool as { name?: string }).name ?? "unknown")
        : "unknown";
    this._execution.recordToolCall({ name: `langchain:${toolName}` });
  }

  async handleChainError(): Promise<void> {
    // no-op: don't double-account errors; the engine's state is fine
  }
}
